src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -71,99 +83,280 @@ public long BytesWritten
+        /// <exception cref="ArgumentNullException">
+        /// Thrown when the instance of <see cref="Stream" /> that is passed in is null.
+        /// </exception>
+        public Utf8JsonWriter(Stream utf8Json, JsonWriterOptions options = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+            if (!utf8Json.CanWrite)
+                throw new ArgumentException(SR.StreamNotWritable);
+
+            _stream = utf8Json;
+            _arrayBufferWriter = new ArrayBufferWriter<byte>();
+            _output = _arrayBufferWriter;
+
+            BytesPending = default;
+            BytesCommitted = default;
+            _memory = default;
+
+            _inObject = default;
+            _isNotPrimitive = default;
+            _tokenType = default;
+            _currentDepth = default;
+            Options = options;
+
+            // Only allocate if the user writes a JSON payload beyond the depth that the _allocationFreeContainer can handle.
+            // This way we avoid allocations in the common, default cases, and allocate lazily.
+            _bitStack = default;
+        }
+
+        /// <summary>
+        /// Resets the <see cref="Utf8JsonWriter"/> internal state so that it can be re-used.
+        /// </summary>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> will continue to use the original writer options
+        /// and the original output as the destination (either <see cref="IBufferWriter{Byte}" /> or <see cref="Stream" />).
+        /// </remarks>
+        public void Reset()
+        {
+            if (_arrayBufferWriter != null)
+            {
+                _arrayBufferWriter.Clear();
+            }
+            ResetHelper();
+        }
+
+        /// <summary>
+        /// Resets the <see cref="Utf8JsonWriter"/> internal state so that it can be re-used with the new instance of <see cref="Stream" />.
+        /// </summary>
+        /// <param name="utf8Json">An instance of <see cref="Stream" /> used as a destination for writing JSON text into.</param>
         /// <remarks>
-        /// Unlike the <see cref="Utf8JsonWriter"/>, which is a ref struct, the state can survive
-        /// across async/await boundaries and hence this type is required to provide support for reading
-        /// in more data asynchronously before continuing with a new instance of the <see cref="Utf8JsonWriter"/>.
+        /// The <see cref="Utf8JsonWriter"/> will continue to use the original writer options
+        /// but now write to the passed in <see cref="Stream" /> as the new destination.
         /// </remarks>
-        public JsonWriterState GetCurrentState()
+        /// <exception cref="ArgumentNullException">
+        /// Thrown when the instance of <see cref="Stream" /> that is passed in is null.
+        /// </exception>
+        public void Reset(Stream utf8Json)
```

Comment:
I don't guard against passing in the same instance of the stream. Should I?

Reply:
No, it's not technically wrong.